doc/dev/mon-osdmap-prune.rst

```diff
@@ -0,0 +1,289 @@
+steps to resolve this particular corner case because we don't think it
+will nefariously affect the pruning process.
+
+With our maps pinned, we will now be able to remove the maps in-between.
+
+We could do the removal in one go, but we have no idea how long that would
+take. Therefore, we will perform several iterations, removing at most
+``mon_osdmap_full_prune_txsize`` osdmaps per iteration.
+
+In the end, our on-disk map sequence will look similar to::
+
+    ------------------------------------------
+    |1|10|20|30|..|49500|49501|..|49999|50000|
+    ------------------------------------------
+     ^ first                           last ^
+
+
+Because we are not pruning all versions in one go, we need to keep state
+about how far along on our pruning we are. With that in mind, we have
+created a data structure, ``osdmap_manifest_t``, that holds the set of pinned
+maps and the last pruned epoch: ::
+
+    struct osdmap_manifest_t:
+        set<version_t> pinned;
+        version_t      last_pruned;
+
+The osdmap manifest will be written to disk the first time we prune maps.
+We ensure that the manifest is written to disk with every transaction
+resulting from a pruning round, so that we always have an up-to-date
+manifest.
+
+We also rely on the manifest existing on disk to ascertain whether we are in
+the middle of pruning, or have pruned in the past. If we have pruned at
+least once, we will require the manifest so that we know which maps have
+been pinned, in order to rebuild osdmaps from incremental maps.
+Additionally, we need to know which maps we have pinned during osdmap
+trimming, but we will discuss that later; for now, let us focus on the
+prunning algorithm.
+
+For simplicity, we have decided that we shall not further along the pruning
+interval if a pruning is yet to be finished.
+
+Upon starting a full osdmap prune, we will pin all the maps we need (as
+discussed before), and will keep the pinned map vector in the osdmap manifest.
+The manifest will also be initiated with ``last_pruned`` at zero, denoting
+that there hasn't been a single version over ``first_committed`` that has been
+removed; in reality, these should be the criteria when ascertaining whether
+pruning has been started, not started, or has finished:
+
+* We will have not started pruning if ::
+  
+  invariant: first_committed > 1
+  last_pruned < first_committed
+
+* We will have started pruning if ::
+  
+  invariant: first_committed > 0
+  invariant: !pinned.empty())
+  invariant: pinned.count(first_committed) == 1
+  invariant: last_pruned > 0
+  invariant: last_pruned < last_committed
+  
+  precond:  last_pruned > first_committed AND
+  last_pruned < pinned.last() AND
+  !finished_pruning
+  
+  postcond: last_pruned > first_committed AND
+  last_pruned < pinned.last() AND
+  upper_pinned(last_pruned) C pinned AND
+  lower_pinned(last_pruned) C pinned AND
+  upper_pinned(last_pruned) != lower_pinned(last_pruned)
+
```

Comment:
indentation was messed up for these last two blocks. I'll address it before merging.